kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/AsyncTransformer.java

```diff
@@ -0,0 +1,96 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.function.Supplier;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+@Slf4j
+@SuppressWarnings("UnstableApiUsage")
+public abstract class AsyncTransformer<K, V, KOUT, VOUT>
+    implements Transformer<K, V, KeyValue<KOUT, VOUT>> {
+  private final Executor executor;
+  private final BlockingQueue<CompletableFuture<List<KeyValue<KOUT, VOUT>>>> pendingFutures;
+  private final RateLimiter rateLimiter;
+  private ProcessorContext context;
+
+  // TODO: configurable executor - supplier pattern. This enables to use a common thread-pool for
+  // all stream tasks
+  public AsyncTransformer(int concurrency, int maxBatchSize, Duration flushInterval) {
+    ThreadFactory threadFactory =
+        new ThreadFactoryBuilder()
+            .setNameFormat("async-transformer-pool-%d")
+            .setDaemon(true)
+            .build();
+    this.executor = Executors.newFixedThreadPool(concurrency, threadFactory);
+    this.pendingFutures = new ArrayBlockingQueue<>(maxBatchSize);
+    this.rateLimiter = RateLimiter.create(1.0 / flushInterval.toSeconds());
+  }
+
+  public AsyncTransformer(
+      Supplier<Executor> executorSupplier, int maxBatchSize, Duration flushInterval) {
+    this.executor = executorSupplier.get();
+    this.pendingFutures = new ArrayBlockingQueue<>(maxBatchSize);
+    this.rateLimiter = RateLimiter.create(1.0 / flushInterval.toSeconds());
+  }
+
+  @Override
+  public void init(ProcessorContext context) {
+    this.context = context;
+    doInit(context.appConfigs());
+  }
+
+  protected abstract void doInit(Map<String, Object> appConfigs);
+
+  public abstract List<KeyValue<KOUT, VOUT>> asyncTransform(K key, V value);
+
+  @SneakyThrows
+  @Override
+  public final KeyValue<KOUT, VOUT> transform(K key, V value) {
+
+    CompletableFuture<List<KeyValue<KOUT, VOUT>>> future =
+        CompletableFuture.supplyAsync(() -> asyncTransform(key, value), executor);
+    // with put, thread gets blocked when queue is full. queue consumer runs in this same thread.
+    pendingFutures.put(future);
+
+    // Flush based on size or time - whichever occurs first
+    // once the queue is full, flush the queue.
+    if (pendingFutures.remainingCapacity() == 0 || rateLimiter.tryAcquire()) {
+      log.debug("flush start - queue size before flush: {}", pendingFutures.size());
+      processResults();
+      log.debug("flush end - queue size after flush: {}", pendingFutures.size());
+    }
+    return null;
+  }
+
+  @Override
+  public void close() {
+    // no-op
+  }
+
+  @SneakyThrows
+  private void processResults() {
+    while (!pendingFutures.isEmpty()) {
```

Comment:
How does this ensure in-order processing? 
msg1/offset1 -- future1
msg2/offset2-- future2
msg3/offset3 -- future3

say future3 completes first and future2 and future1 is not yet done, what offset will get commited?

Reply:
Both the PR and my suggestion would ensure in order processing, but the offset behavior would be a bit different.

In the current PR, future 3 could complete first but it doesn't matter, we don't check it until future 1 has completed and its result forwarded, then 2 and then finally 3. Once 3 (or whichever is the last of the batch) is forwarded, offset 3 would be committed. So the forwarding happens in order, the commit happens at the end of the batch.

In my proposal in the above comment, the behavior change I'm going for is that we wait for the whole batch of futures to complete before forwarding. Again in order, but this way if future 3 were to error (which means we'd never commit), we haven't forwarded 1 and 2. In the PR version, those two would have been forwarded and uncommitted, so would have to be re-sent.

Reply:
Discussed this offline with Laxman, if this is used with AnomalyEvent component, we might have duplicate events on resent scenarios. But, it seems we have at least once consistency in our system.

Reply:
Not sure why we didn't do this - this was an easy fix to avoid the duplicate sends on errors?

Reply:
- @mihirgt This is in order processing only. 

-  @aaron-steinfeld: we discussed in detail offline. I thought this was clarified. its harder for me to explain kafka streams context in comments. we can implement this. but this has very limited scope of preventing duplicates. also, the duplicates extent and probability of duplicates is still the same with async transformer as well. i feel this adds no value.  if you still insist, i can add this logic.

Reply:
Last I recall you had said you were going to try it out, but maybe there was another thread I missed or am forgetting. Anyway, I'm not insisting that you add the logic if you feel it adds no value, but I do think we should sync so we can get on the same page whether it does or not, then add the rationale here.

>  but this has very limited scope of preventing duplicates. also, the duplicates extent and probability of duplicates is still the same with async transformer as well. 

This is the part I'm not understanding. If we forward each as it completes, then a later one fails before we commit, then we get duplicates. If we collect the whole batch first, we know it's successful so we should not except in the rare case where there's unavailability or a shutdown in the midst of forwarding the already-calculated results. Is this not the case, or are there common scenarios where we get duplicates that means this improvement is worthwhile? Or are we concerned that will slow down processing because we're not forwarding messages continuously, but in discrete chunks? Or is it something else?

Reply:
> If we forward each as it completes, then a later one fails before we commit, then we get duplicates.
Keep aside async for a moment. That's exactly how kafka streams plain vanilla transformer works. That's why I was saying 
### extent and probability of duplicates with async is still same (not worse/better)

> If we collect the whole batch first, we know it's successful
Downstream, next node the in topology will soon process and forward immediately. This neutralises the logic we add to prevent duplicates. 

>are there common scenarios where we get duplicates
Yes. any node(transformer/mapper/processor) failure downstream will still duplicate the entire batch. The probability of other transformer/processors failure is same as the async transformer failure.

>Or are we concerned that will slow down
No. not concerned about slow down at all. I am just trying to say, this additional logic has a very limited scope in preventing duplicates.

Reply:
> Keep aside async for a moment. That's exactly how kafka streams plain vanilla transformer works. That's why I was saying
> extent and probability of duplicates with async is still same (not worse/better)

That's fair and makes sense. This would not be different than vanilla other than that we're controlling the batch sizes differently (not sure how they compare if they're bigger or smaller). My thinking was that since we're doing it manually anyway, gathering up all the responses so when we know to commit, we've got all the data we need to buffer then send so why not improve on it?

> Downstream, next node the in topology will soon process and forward immediately. This neutralises the logic we add to prevent duplicates.

The next node could still create duplicates if it's not using this async transformer, but I'm not sure I'm following how that neutralizes it? A single node creating duplicates would result in less duplicates than multiple nodes creating them.

> Yes. any node(transformer/mapper/processor) failure downstream will still duplicate the entire batch. The probability of other transformer/processors failure is same as the async transformer failure.

Maybe I'm misunderstanding, so let me illustrate how I was imagining it and hopefully you can help clarify.

Imagine you've got async node A, and a vanilla node B downstream. Now imagine that while processing the second message in node A, we get a failure.

In the vanilla scenario, node A would send the first message, fail to commit, restart and send the first message again - we'd repeatedly be sending that first message until whatever caused the issue processing it was resolved (or maybe it's transient and we only send it one extra time). Node B would receive 2 or more copies, filling up its batch and then sending. If node B doesn't fail, it forwards on the output produced from those 2+ messages, along with any others that fit in the batch. If node B does fail, it'd duplicate again and now wouldn't we potentially get (depending on where it failed) duplicates of the duplicates.

In the async scenario, because there's less places duplication can happen, there's less overall duplicates. It wouldn't be zero, but each node that can duplicate is multiplicative down the pipeline I thought.

> Any failure in kafka streams mostly duplicates entire batch. I don't want to wrongly claim with lots of disclaimers that async can prevent duplicates better.

It's not really the async part doing it, I was thinking that because it's async there's less of a penalty to waiting for the entire batch. In a default sync impl, I don't think kafka can really make that assumption

Anyway, like I said - more than anything, I'm trying to learn here. If my mental is wrong or incomplete, then I'd love to update it.

Reply:
>Maybe I'm misunderstanding, so let me illustrate

Yes. I feel there is some misunderstanding.

Stream thread does something like this in the following pseudo code. Ignore the nits and don't look for the perfection of this pseudo code. This is just for understanding.

```
// fetch a batch (Ex: 100) of events (say, msgs) from input topic partition 
for (msg in msgs) {
  msg = Node-A->process(msg)
  msg = Node-B->process(msg)
  msg = Node-C->process(msg)
  ...
  sink.write(msg) // writes to output topic
}

if (it's time to commit) {
  commit the latest offset (100, in this case)
}
```

As you can see above, **each message is passed through all the eligible nodes of topology one after the other**. There is no batching at each node level in kafka streams. Here, if there is a problem with msg-30 at a node (say, node-D), then topology will fail, pod will crash and it will try reprocess all the msgs (msg-1 to msg-100 causing msg-1 to msg-30 duplicated at the sink) on restart. 

If its still not clear I prefer to explain offline and you can help me summarise here.

Reply:
Appreciate the illustration, thank you. I believe I'm following now. I had thought that when you were saying node, you were talking about the broader kafka topology (e.g. a downstream consumer), but it sounds you mean within the single consumer's topology if any node were to fail, the whole batch would have to restart regardless of the commit behavior of any one particular node.

Given that I'm interested in understanding how the commit behavior of individual nodes interact with the overall consumer group's commit behavior (for my own knowledge), but I see your point why that would de-emphasize the commit behavior of a particular node. That is, unless the vast majority of errors originate in nodes that would be moving to async.

Anyway, maybe for now it would be helpful if we could set up a general kafka or kafka streams knowledge sharing session to help myself and others to understand the nuts and bolts of how the stream topology works, what all the various configs are used for and when to change them etc?